<!DOCTYPE html><html lang="fr"><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width, initial-scale=1"/><link rel="preload" as="image" href="/images/logo.svg"/><link rel="stylesheet" href="https://insertafter.com/_next/static/css/655a273f6fe37552.css" crossorigin="" data-precedence="next"/><link rel="stylesheet" href="https://insertafter.com/_next/static/css/cd8d1ff5c688cf4d.css" crossorigin="" data-precedence="next"/><link rel="preload" as="script" fetchPriority="low" href="https://insertafter.com/_next/static/chunks/webpack-bc3269b1fdd7d5a8.js" crossorigin=""/><script src="https://insertafter.com/_next/static/chunks/fd9d1056-5bc70f044f3cb802.js" async="" crossorigin=""></script><script src="https://insertafter.com/_next/static/chunks/938-bf741a2eb8fd4828.js" async="" crossorigin=""></script><script src="https://insertafter.com/_next/static/chunks/main-app-2e6280b5273157a5.js" async="" crossorigin=""></script><script src="https://insertafter.com/_next/static/chunks/176-0a03be62efe5672b.js" async=""></script><script src="https://insertafter.com/_next/static/chunks/app/(withLocale)/%5Blocale%5D/layout-3730889d272dc34b.js" async=""></script><script src="https://insertafter.com/_next/static/chunks/790-a194f74901d33197.js" async=""></script><script src="https://insertafter.com/_next/static/chunks/app/(withLocale)/%5Blocale%5D/%5B%5B...slug%5D%5D/page-992952bac383cfc2.js" async=""></script><meta name="theme-color" content="#3e78b2"/><title>Gestion des erreurs en JavaScript - Nicolas Froidure</title><meta name="description" content="GÃ©rer les erreurs en JavaScript peut paraÃ®tre simple, mais il n&#x27;est pas toujours Ã©vident de choisir entre exceptions, retours d&#x27;erreurs et callbacks d&#x27;erreurs."/><link rel="author" href="https://insertafter.com"/><meta name="author" content="Nicolas Froidure"/><meta name="robots" content="index, follow"/><link rel="canonical" href="https://insertafter.com/fr/blog/erreurs_javascript"/><meta property="og:title" content="Gestion des erreurs en JavaScript - Nicolas Froidure"/><meta property="og:description" content="GÃ©rer les erreurs en JavaScript peut paraÃ®tre simple, mais il n&#x27;est pas toujours Ã©vident de choisir entre exceptions, retours d&#x27;erreurs et callbacks d&#x27;erreurs."/><meta property="og:url" content="https://insertafter.com/fr/blog/erreurs_javascript"/><meta property="og:site_name" content="Nicolas Froidure"/><meta property="og:locale" content="fr_FR"/><meta property="og:image:alt" content="BanniÃ¨re du site"/><meta property="og:image" content="https://insertafter.com/images/banner.png"/><meta property="og:type" content="article"/><meta name="twitter:card" content="summary"/><meta name="twitter:site" content="@nfroidure"/><meta name="twitter:creator" content="@nfroidure"/><meta name="twitter:title" content="Gestion des erreurs en JavaScript - Nicolas Froidure"/><meta name="twitter:description" content="GÃ©rer les erreurs en JavaScript peut paraÃ®tre simple, mais il n&#x27;est pas toujours Ã©vident de choisir entre exceptions, retours d&#x27;erreurs et callbacks d&#x27;erreurs."/><meta name="twitter:image:alt" content="BanniÃ¨re du site"/><meta name="twitter:image" content="https://insertafter.com/images/banner.png"/><link rel="icon" href="/images/favicon.svg" type="image/svg+xml" sizes="any"/><link rel="icon" href="/images/favicon-16.png" type="image/png" sizes="16x16"/><link rel="icon" href="/images/favicon-128.png" type="image/png" sizes="128x128"/><script src="https://insertafter.com/_next/static/chunks/polyfills-c67a75d1b6f99dc8.js" crossorigin="" noModule=""></script></head><body class="layout_body__jFZuu"><header class="header_iaHeader__4vCMA"><div><h1><a href="/fr" title="Retour Ã  lâ€™accueil" class="header_iaHeaderLogo__0uNr1"><img src="/images/logo.svg" alt="Logo dâ€™InsertAfter"/></a></h1></div></header><nav class="menu_iaMenu__B_rtj"><ul class="menu_iaMenuBody__0Zveq"><li class="menu_iaMenuItem__rOch4"><a href="/fr" title="Retour Ã  lâ€™accueil" class="">Accueil</a></li><li class="menu_iaMenuItem__rOch4"><a href="/fr/blog" title="Retour Ã  lâ€™index du blog" class="">Blog</a></li><li class="menu_iaMenuItem__rOch4"><a href="/fr/projets" title="DÃ©couvrir mes divers projets" class="">Projets</a></li><li class="menu_iaMenuItem__rOch4"><a href="/fr/a_propos" title="En savoir plus Ã  mon sujet" class="">Ã€ propos</a></li></ul></nav><nav class="languages_iaLanguages__K2J3W"><ul><li class="languages_iaLanguagesFlag__ruQ4j languages_iaLanguagesFlagEn__angfJ languages_selected__w1dKv"><a href="/en"><span>en</span></a></li><li class="languages_iaLanguagesFlag__ruQ4j languages_iaLanguagesFlagFr__Us5Jb"><a href="/fr"><span>fr</span></a></li></ul></nav><main class="layout_iaMain__6s2gl"><div><section class="contentBlock_root__802UK"><h1 class="h1_root__VPXQG"><span>Gestion des erreurs en JavaScript</span></h1><p class="p_root__SHE_W"><span>GÃ©rer les erreurs en JavaScript peut paraÃ®tre simple, mais il n&#x27;est pas toujours Ã©vident de choisir entre exceptions, retours d&#x27;erreurs et callbacks d&#x27;erreurs.</span></p><p class="p_root__SHE_W"><span>Le dernier Ch&#x27;tiJS a dÃ©cidÃ©ment Ã©tÃ© la source de nombreuses rÃ©flexions. L&#x27;une d&#x27;entre elle concernait les erreurs. Tout est parti d&#x27;une assertion simpleÂ : les promises doivent-elles attraper les exceptions et les transformer en erreurs rÃ©cupÃ©rables dans un then ou les laisser remonter la pile d&#x27;exÃ©cutionÂ ?</span></p><p class="p_root__SHE_W"><span>C&#x27;est que fait, par exemple, RSVP.js. La logique du resolver est englobÃ©e dans </span><a class="a_root__WEPS7" title="Voir le code en question" target="_blank" href="https://github.com/tildeio/rsvp.js/blob/277eb5eab912a9c3df8228a2647fcab84374615a/lib/rsvp/promise.js#L47"><span>un bloc try...catch</span></a><span> qui transforme toute exception en erreur qui fait Ã©chouer la promise. Dans </span><a class="a_root__WEPS7" title="Voir l&#x27;article concernant cette implÃ©mentation" target="_self" href="promises%5Fjavascript%5Fouvrons%5Fcapot.html"><span>mon implÃ©mentation des promises</span></a><span>, j&#x27;ai choisi de ne pas utiliser cette approche afin de pouvoir dÃ©cider, au cas par cas, d&#x27;attraper ou non une exception envoyÃ©e par le resolver d&#x27;une promise.</span></p><h2 class="h2_root__Y02lX"><span class="anchored_root__HWi6K"><span>Typologie des erreurs</span> <small><a class="anchored_icon__N5Sem" id="typologie-des-erreurs" title="Lien vers cette section" href="#typologie-des-erreurs"><span>ğŸ”—</span></a></small></span></h2><p class="p_root__SHE_W"><span>Avant d&#x27;aller plus loin, je voudrais revenir sur la typologie des erreurs que l&#x27;on peut rencontrer au cours de l&#x27;exÃ©cution d&#x27;un programme. Pour moi, il existe trois classes d&#x27;erreursÂ :</span></p><ul class="ul_ul__qh0e9"><li class="li_root__oY9Dv"><p class="p_root__SHE_W"><span>les erreurs fatalesÂ : ce sont des erreurs qui rÃ©sultent d&#x27;une erreur de programmation. Ces derniÃ¨res ont la particularitÃ© de ne pas avoir Ã©tÃ© prÃ©vues par le dÃ©veloppeur puisqu&#x27;elles rÃ©sultent d&#x27;un bug. Elles laissent donc le programme dans un Ã©tat indÃ©terminÃ©. On pourrait comparer cette erreur aux erreurs HTTP 500+ ;</span></p></li><li class="li_root__oY9Dv"><p class="p_root__SHE_W"><span>les erreurs d&#x27;utilisationÂ : elles sont causÃ©es par une mauvaise utilisation du programme (mauvaise saisies de l&#x27;utilisateur principalement) ;</span></p></li><li class="li_root__oY9Dv"><p class="p_root__SHE_W"><span>les erreurs d&#x27;interactionÂ : un programme interagit frÃ©quemment avec d&#x27;autres composants (le rÃ©seau, le matÃ©riel etc...). Il arrive qu&#x27;une interaction se termine de faÃ§on imprÃ©vue ou incohÃ©rente. Un programme se doit d&#x27;anticiper ces erreurs et de les gÃ©rer au mieux.</span></p></li></ul><p class="p_root__SHE_W"><span>Attention, certaines erreurs d&#x27;interaction peuvent Ãªtre dues Ã  une erreur fatale dans un programme avec lequel votre code interagit. C&#x27;est le cas quand une requÃªte XHR renvoie une erreur 500. Il faut donc veiller Ã  ne pas confondre erreurs d&#x27;interactions et erreurs fatales. Pour ce faire, il faut bien dÃ©limiter les erreurs qui sont dues Ã  votre code.</span></p><p class="p_root__SHE_W"><span>On pourrait considÃ©rer que les erreurs d&#x27;utilisation sont des erreurs d&#x27;interaction avec l&#x27;utilisateur. D&#x27;ailleurs, dans l&#x27;absolu, l&#x27;interaction avec des systÃ¨me tiers peut aussi induire des entrÃ©es qui nÃ©cessitent d&#x27;Ãªtre filtrÃ©es de la mÃªme faÃ§on que les entrÃ©es utilisateur. Je fais nÃ©anmoins la diffÃ©rence ici car la rÃ©ponse Ã  ces erreurs sera lÃ©gÃ¨rement diffÃ©rente.</span></p><h2 class="h2_root__Y02lX"><span class="anchored_root__HWi6K"><span>La gestion d&#x27;erreur</span> <small><a class="anchored_icon__N5Sem" id="la-gestion-d-erreur" title="Lien vers cette section" href="#la-gestion-d-erreur"><span>ğŸ”—</span></a></small></span></h2><p class="p_root__SHE_W"><span>J&#x27;imagine que ces types vous Ã©voquent certaines stratÃ©gies pour leur gestion. En JavaScript, on compte divers type de gestion d&#x27;erreurÂ :</span></p><ul class="ul_ul__qh0e9"><li class="li_root__oY9Dv"><p class="p_root__SHE_W"><span>les exceptionsÂ : elles permettent Ã  une erreur de remonter la pile d&#x27;exÃ©cution. On peut les attraper Ã  tout moment Ã  l&#x27;aide d&#x27;un bloc </span><code class="code_root__5JLV_">try...catch</code><span>. Si elles ne sont pas attrapÃ©es, les exceptions finissent par interrompre l&#x27;execution de votre JavaScript (plus prÃ©cisÃ©ment de la pile d&#x27;exÃ©cution JavaScript).</span></p></li><li class="li_root__oY9Dv"><p class="p_root__SHE_W"><span>les retours d&#x27;erreurÂ : il s&#x27;agit d&#x27;exploiter le retour de fonction pour indiquer qu&#x27;une erreur a eu lieu. On peut considÃ©rer que si </span><code class="code_root__5JLV_">parseInt</code><span> retourne </span><code class="code_root__5JLV_">NaN</code><span> alors on a affaire Ã  une erreur. Cette approche est trÃ¨s utilisÃ©e en C pour les appels systÃ¨mes.</span></p></li><li class="li_root__oY9Dv"><p class="p_root__SHE_W"><span>l&#x27;approche via une variable globaleÂ : trÃ¨s prÃ©sente en C Ã©galement avec la variable </span><code class="code_root__5JLV_">errno</code><span>. En JavaScript, on Ã©vite les variables globales, on utilisera donc pas cette approche.</span></p></li><li class="li_root__oY9Dv"><p class="p_root__SHE_W"><span>l&#x27;approche via callback d&#x27;erreurÂ : ils sont principalement utilisÃ©s pour les opÃ©rations asynchrones, bien que rien n&#x27;empÃªche leur utilisation dans du code synchrone. C&#x27;est aussi l&#x27;approche qui est utilisÃ©e pour les promises. On passe un callback d&#x27;erreur Ã  la mÃ©thode then. La diffÃ©rence est que l&#x27;on passe autant de callbacks d&#x27;erreur qu&#x27;on le souhaite.</span></p></li></ul><h2 class="h2_root__Y02lX"><span class="anchored_root__HWi6K"><span>Comment gÃ©rer chaque type d&#x27;erreurÂ ?</span> <small><a class="anchored_icon__N5Sem" id="comment-gerer-chaque-type-d-erreur" title="Lien vers cette section" href="#comment-gerer-chaque-type-d-erreur"><span>ğŸ”—</span></a></small></span></h2><h3 class="h3_root__kVaqq"><span class="anchored_root__HWi6K"><span>Les erreurs fatales</span> <small><a class="anchored_icon__N5Sem" id="les-erreurs-fatales" title="Lien vers cette section" href="#les-erreurs-fatales"><span>ğŸ”—</span></a></small></span></h3><p class="p_root__SHE_W"><span>Les exceptions sont l&#x27;arme ultime pour gÃ©rer les erreurs fatales. En effet, puisque les erreurs fatales sont des erreurs de programmation, continuer l&#x27;exÃ©cution du programme est risquÃ©. Cela pourrait avoir des rÃ©percussions dÃ©sastreuses.</span></p><p class="p_root__SHE_W"><span>JavaScript utilise intensivement les exceptions. Les erreurs de syntaxe, les mauvaises utilisations de son API native sont signalÃ©es dans la grande majoritÃ© des cas par des exceptions. C&#x27;est pourquoi je vous enjoins Ã  fait de mÃªme.</span></p><p class="p_root__SHE_W"><span>La plupart du temps, il s&#x27;agit de lever une exception si l&#x27;une de vos fonctions/l&#x27;un de vos constructeurs reÃ§oit en paramÃ¨tre des arguments erronÃ©sÂ :</span></p><div class="code_root__5JLV_"><code><pre>function repeat(n) {
  if (n &lt; 0) throw new Error(&quot;Bad n value.&quot;);
  //...
}</pre></code></div><p class="p_root__SHE_W"><span>Demander une rÃ©pÃ©tition d&#x27;une opÃ©ration avec une nombre nÃ©gatif de rÃ©pÃ©tition n&#x27;a pas de sens. Lancer une exception permet d&#x27;interrompre l&#x27;exÃ©cution et de permettre de dÃ©boguer facilement l&#x27;erreur dans le contexte oÃ¹ celle-ci est apparue.</span></p><p class="p_root__SHE_W"><span>En effet, tous les outils pour dÃ©veloppeurs permettent de placer un point d&#x27;arrÃªt Ã  la levÃ©e d&#x27;une exception afin de permettre de visualiser les variables Ã  portÃ©es, la pile d&#x27;exÃ©cution et bien d&#x27;autres informations trÃ¨s utiles.</span></p><h3 class="h3_root__kVaqq"><span class="anchored_root__HWi6K"><span>Les erreurs d&#x27;utilisation</span> <small><a class="anchored_icon__N5Sem" id="les-erreurs-d-utilisation" title="Lien vers cette section" href="#les-erreurs-d-utilisation"><span>ğŸ”—</span></a></small></span></h3><p class="p_root__SHE_W"><span>Les erreurs d&#x27;utilisation sont dues Ã  l&#x27;utilisateur. La logique veut donc que l&#x27;on traite ces erreurs en signalant celles-cis Ã  l&#x27;utilisateur. C&#x27;est donc une rÃ©ponse par le biais de l&#x27;interface utilisateur qui sera privilÃ©giÃ©e.</span></p><p class="p_root__SHE_W"><span>Cependant, pour activer cette rÃ©ponse, il faut en premier lieu la dÃ©tecter. Cette dÃ©tection se fait indiffÃ©remment par retour de fonction ou par callback d&#x27;erreur. D&#x27;ailleurs il est trÃ¨s possible que vous deviez mixer ces deux techniques pour bÃ¢tir une propagation d&#x27;erreur dans la pile d&#x27;exÃ©cution et annuler certaines actions suite Ã  une erreur.</span></p><p class="p_root__SHE_W"><span>L&#x27;idÃ©e Ã  retenir est d&#x27;Ã©viter Ã  tout prix de gÃ©rer les erreurs des utilisateurs Ã  l&#x27;aide d&#x27;exceptions. En effet, vous prenez le risque de mettre fin Ã  l&#x27;exÃ©cution pour une erreur normale dans le cycle de vie d&#x27;une application. Les utilisateurs se trompent souvent.</span></p><p class="p_root__SHE_W"><span>On prendra donc soin de filtrer un entier obtenu grÃ¢ce Ã  une saisie de l&#x27;utilisateur avant de l&#x27;utiliser comme argument de la fonction </span><code class="code_root__5JLV_">repeat</code><span> dÃ©finie ci-dessus.</span></p><h3 class="h3_root__kVaqq"><span class="anchored_root__HWi6K"><span>Les erreurs d&#x27;interaction</span> <small><a class="anchored_icon__N5Sem" id="les-erreurs-d-interaction" title="Lien vers cette section" href="#les-erreurs-d-interaction"><span>ğŸ”—</span></a></small></span></h3><p class="p_root__SHE_W"><span>Lorsque votre code interagit avec des Ã©lÃ©ments extÃ©rieurs, qu&#x27;il s&#x27;agisse de serveurs distant ou du hardware, le bon sens veut que cela se fasse de maniÃ¨re asynchrone. Les callbacks d&#x27;erreur sont donc indiquÃ©s pour remplir le job. C&#x27;est d&#x27;ailleurs de cette faÃ§on que l&#x27;API JavaScript vous informe sur le succÃ¨s ou l&#x27;Ã©chec d&#x27;une opÃ©ration asynchrone.</span></p><p class="p_root__SHE_W"><span>Ce n&#x27;est malheureusement pas toujours le cas. Par exemple, une requÃªte XHR synchrone qui Ã©choue Ã  cause d&#x27;un problÃ¨me rÃ©seau lance une exception. Dans ce cas prÃ©cis, mieux vaut Ãªtre au courant afin de placer le bloc </span><code class="code_root__5JLV_">try...catch</code><span> qui vous permettra de gÃ©rer ce cas sans mettre en pÃ©ril l&#x27;exÃ©cution de votre JavaScript.</span></p><p class="p_root__SHE_W"><span>Vous vous dÃ®tes peut-Ãªtre qu&#x27;il peut arriver qu&#x27;une erreur de programmation mÃ¨ne Ã  la mauvaise utilisation d&#x27;une API asynchrone et que dans ce cas, un callback d&#x27;erreur sera appelÃ© pour une erreur fatale. Ce que j&#x27;ai totalement contre indiquÃ© tout Ã  l&#x27;heure.</span></p><p class="p_root__SHE_W"><span>Pour le cas d&#x27;une API bien pensÃ©e, cela ne devrait pas Ãªtre le cas. Une API asynchrone doit dÃ©tecter toute mauvaise utilisation et lancer une exception au moment de son lancement. Par exemple, si vous lancez une requÃªte XHR vers une origine diffÃ©rente de votre site, c&#x27;est une erreur de programmation. Une exception doit Ãªtre lancÃ©e immÃ©diatement.</span></p><p class="p_root__SHE_W"><span>L&#x27;astuce pour Ã©viter tout effet de bord est de lancer l&#x27;opÃ©ration asynchrone avant d&#x27;assigner les callbacks ou Ã©couteurs d&#x27;Ã©vÃ¨nements. De cette faÃ§on vous Ãªtes certains que ces derniers ne seront jamais appelÃ©s en cas de levÃ©e d&#x27;une exception.</span></p><h2 class="h2_root__Y02lX"><span class="anchored_root__HWi6K"><span>Retour sur les promises</span> <small><a class="anchored_icon__N5Sem" id="retour-sur-les-promises" title="Lien vers cette section" href="#retour-sur-les-promises"><span>ğŸ”—</span></a></small></span></h2><p class="p_root__SHE_W"><span>Avec cette toile de fond en tÃªte, il apparaÃ®t donc Ã©vident que les exceptions ne doivent pas Ãªtre attrapÃ©es au sein des promises. En effet, au delÃ  de l&#x27;avantage d&#x27;une gestion au cas par cas, on voit aussi que cela pourrait compromettre la dÃ©tection d&#x27;une erreur fatale causant la continuation de l&#x27;exÃ©cution Ã  partir d&#x27;un Ã©tat non-prÃ©vu donc indÃ©terminÃ©.</span></p><p class="p_root__SHE_W"><span>On pourrait penser Ã©pargner l&#x27;utilisateur d&#x27;erreurs fatales qui en dÃ©finitive mettent fin Ã  l&#x27;exÃ©cution, mais au final, on rend les erreurs de programmation plus difficiles Ã  dÃ©tecter et les bugs plus sournois.</span></p><p class="p_root__SHE_W"><span>Il faut comprendre que les erreurs fatales doivent Ãªtre exceptionnelles. Un programme qui arrive en production doit Ãªtre dÃ©nuÃ© d&#x27;erreurs fatales dans l&#x27;absolu. Malheureusement, dans le monde rÃ©el, ces erreurs peuvent survenir en production.</span></p><p class="p_root__SHE_W"><span>Devons nous laisser les utilisateurs face Ã  une application qui ne fonctionne plusÂ ? SÃ»rement pas. L&#x27;approche correcte pour gÃ©rer les erreurs fatales est finalement assez simple. Il suffit d&#x27;utiliser un callback d&#x27;erreur gÃ©nÃ©rique, d&#x27;informer l&#x27;utilisateur d&#x27;une erreur anormale et de redÃ©marrer l&#x27;application. On peut aussi imaginer rÃ©cupÃ©rer ces erreurs fatales grÃ¢ce Ã  une requÃªte XHR et de crÃ©er directement une issue sur GitHub. Un simple exempleÂ :</span></p><div class="code_root__5JLV_"><code><pre>window.onerror = function (msg, url, line) {
  //demande de dÃ©tails
  var details = prompt(
    &quot;Une erreur fatale a eu lieu, vous pouvez donner quelques dÃ©tails pour aider le webmasterÂ :&quot;,
  );
  // envoi des dÃ©tails via XHR puis rafraÃ®chissement de la page
  document.location.refresh();
};</pre></code></div><p class="p_root__SHE_W"><span>Ici, j&#x27;ai volontairement simplifiÃ©. Dans la pratique, on placera un timeout pour laisser le temps Ã  toutes les erreurs consÃ©cutives d&#x27;Ãªtre levÃ©es avant de demander du feedback et de redÃ©marrer l&#x27;application. On peut aussi donner le temps Ã  l&#x27;utilisateur de rÃ©cupÃ©rer certains contenus (un formulaire Ã  moitiÃ© rempli, par exemple) avant le rafraÃ®chissement.</span></p><p class="p_root__SHE_W"><span>L&#x27;avantage de cette approche est qu&#x27;on informe l&#x27;utilisateur d&#x27;un problÃ¨me plutÃ´t que vouloir le lui cacher quitte Ã  ce que ce dernier provoque un effet papillon plus difficile Ã  dÃ©tecter.</span></p><p class="p_root__SHE_W"><span>VoilÃ , j&#x27;espÃ¨re que cet article vous a donnÃ© envie de gÃ©rer les erreurs de vos applications. Je vous propose de lire ce billet sur l&#x27;</span><a class="a_root__WEPS7" title="Voir le billet sur BugMeBack" target="_self" href="bugmeback.html"><span>extension BugMeBack</span></a><span> qui propose une gestion intÃ©ressante des erreurs pour vos applications.</span></p><p class="p_root__SHE_W"><span>Petit ajoutÂ : J&#x27;ai Ã©crit </span><a class="a_root__WEPS7" title="Lire l&#x27;article" target="_self" href="erreurs%5Fjavascript%5Fameliorees.html"><span>un autre article sur les erreurs en JS</span></a><span> mais cette fois avec des exemples concrets.</span></p><p class="p_root__SHE_W">Published at<!-- --> <!-- -->mardi 30 juillet 2013 Ã  15:18:42<!-- -->.</p></section><section class="contentBlock_root__802UK"><aside class="share_root__Qm_RR"><h2 class="h2_root__Y02lX">Commenter et partager</h2><p class="p_root__SHE_W"><a class="a_root__WEPS7" title="Commenter sur Facebook" target="_blank" href="https://www.facebook.com/sharer/sharer.php?u=https%3A%2F%2Finsertafter.com%2Fblog%2Ferreurs_javascript"><span class="share_icon__bi4Q3 share_facebook__JKn8d"></span>Facebook</a> - <a class="a_root__WEPS7" title="Commenter sur Twitter" target="_blank" href="https://twitter.com/intent/tweet/?url=https%3A%2F%2Finsertafter.com%2Fblog%2Ferreurs_javascript&amp;text=Gestion%20des%20erreurs%20en%20JavaScript&amp;via=nfroidure"><span class="share_icon__bi4Q3 "></span>Twitter</a> - <a class="a_root__WEPS7" title="Me rÃ©pondre par mail" target="_self" href="mailto:nicolas.froidure@gmail.com"><span class="share_icon__bi4Q3 share_mail__yvBDK"></span>Courriel</a></p></aside></section><section class="contentBlock_root__802UK"><aside class="page_linkedEntries__jiPuT"><h2 class="h2_root__Y02lX">Articles similaires</h2><p class="p_root__SHE_W">Dans les catÃ©gories<!-- --> <!-- -->JavaScript et Gestion des erreurs<!-- -->.</p><div class="items_entries__5hkLG"><div class="items_entry_item__GWeg7"><h2 class="h2_root__Y02lX items_entry_title__vJ_8h"><a class="a_root__WEPS7 items_entry_link__EtVod" title="Lire cet article de blog" target="_self" href=".//choisir_module_nodejs">Bien choisir ses modules NodeJSÂ : ma mÃ©thode</a></h2><p class="p_root__SHE_W items_entry_description__ZWt9y">Hier Ã  la troisiÃ¨me Ã©dition de ChtiJS, un question a Ã©tÃ© soulevÃ©e sur la confiance que l&#x27;on pouvait avoir dans les modules NodeJS. Je vous dÃ©voile donc ma mÃ©thode pour trancher entre npm install et npm remove ;).<!-- --> <br/><a class="a_root__WEPS7" title="Lire cet article de blog" target="_self" href=".//choisir_module_nodejs">Lire la suite</a></p><p class="p_root__SHE_W items_entry_time__iNrot">Published at<!-- --> <!-- -->vendredi 26 juillet 2013 Ã  16:04:25<!-- -->.</p><div class="items_clear__3fyGk"></div></div><div class="items_entry_item__GWeg7"><h2 class="h2_root__Y02lX items_entry_title__vJ_8h"><a class="a_root__WEPS7 items_entry_link__EtVod" title="Lire cet article de blog" target="_self" href=".//utiliser_requirejs">Astuces pour utiliser pleinement RequireJS</a></h2><p class="p_root__SHE_W items_entry_description__ZWt9y">Pour un projet dont je vous reparlerai bientÃ´t, j&#x27;ai utilisÃ© RequireJS pour la premiÃ¨re fois en production. J&#x27;en profite pour vous faire un petit retour d&#x27;expÃ©rience avec quelques astuces de mon cru ;).<!-- --> <br/><a class="a_root__WEPS7" title="Lire cet article de blog" target="_self" href=".//utiliser_requirejs">Lire la suite</a></p><p class="p_root__SHE_W items_entry_time__iNrot">Published at<!-- --> <!-- -->jeudi 11 juillet 2013 Ã  11:57:45<!-- -->.</p><div class="items_clear__3fyGk"></div></div><div class="items_entry_item__GWeg7"><h2 class="h2_root__Y02lX items_entry_title__vJ_8h"><a class="a_root__WEPS7 items_entry_link__EtVod" title="Lire cet article de blog" target="_self" href=".//promises_javascript_ouvrons_capot">Les promises avec JavaScriptÂ : ouvrons le capotÂ !</a></h2><p class="p_root__SHE_W items_entry_description__ZWt9y">Dans le petit monde de JavaScript les promises font l&#x27;actualitÃ©. Je vous propose de comprendre les promises par l&#x27;implÃ©mentation.<!-- --> <br/><a class="a_root__WEPS7" title="Lire cet article de blog" target="_self" href=".//promises_javascript_ouvrons_capot">Lire la suite</a></p><p class="p_root__SHE_W items_entry_time__iNrot">Published at<!-- --> <!-- -->samedi 8 juin 2013 Ã  12:04:17<!-- -->.</p><div class="items_clear__3fyGk"></div></div></div></aside></section></div></main><aside class="social_iaSocial__kdR1x"><nav><ul><li><a href="https://mastodon.social/@nfroidure" rel="me" title="Me suivre sur Mastodon" class="social_iaSocialMastodon__MTF6o"><span>Mastodon</span></a></li><li><a href="https://bsky.app/profile/nfroidure.bsky.social" rel="me" title="Me suivre sur Bluesky" class="social_iaSocialBluesky__fKW6M"><span>Bluesky</span></a></li><li><a href="https://twitter.com/nfroidure" title="Me suivre sur Twitter" class="social_iaSocialTwitter__9nU6R"><span>Twitter</span></a></li><li><a href="https://github.com/nfroidure" title="Voir mes projets sur GitHub" class="social_iaSocialGithub__vcU5_"><span>GitHub</span></a></li><li><a href="https://www.npmjs.org/~nfroidure" title="Voir mon compte NPM" class="social_iaSocialNpm__6nku5"><span>NPM</span></a></li><li><a href="https://www.linkedin.com/in/nfroidure/" rel="me" title="Voir mon profil LinkedIn" class="social_iaSocialLinkedin__pWACO"><span>LinkedIn</span></a></li><li><a href="https://facebook.com/nicolas.froidure.douai" rel="me" title="Me suivre sur Facebook" class="social_iaSocialFacebook__w4wfH"><span>Facebook</span></a></li><li><a href="/fr/blog/index.atom" title="Sâ€™abonner au flux RSS" class="social_iaSocialFeed__gEQQ0"><span>Flux RSS</span></a></li></ul></nav></aside><footer class="footer_iaFooter__rhlnQ"><p class="footer_iaFooterContent__VZndF">Â© Nicolas Froidure 2012-<!-- -->2024</p></footer><script src="https://insertafter.com/_next/static/chunks/webpack-bc3269b1fdd7d5a8.js" crossorigin="" async=""></script><script>(self.__next_f=self.__next_f||[]).push([0]);self.__next_f.push([2,null])</script><script>self.__next_f.push([1,"0:\"$L1\"\n"])</script><script>self.__next_f.push([1,"2:HL[\"https://insertafter.com/_next/static/css/655a273f6fe37552.css\",\"style\",{\"crossOrigin\":\"\"}]\n3:HL[\"https://insertafter.com/_next/static/css/cd8d1ff5c688cf4d.css\",\"style\",{\"crossOrigin\":\"\"}]\n"])</script><script>self.__next_f.push([1,"4:I[7690,[],\"\"]\n7:I[5613,[],\"\"]\na:I[1778,[],\"\"]\n11:I[8955,[],\"\"]\n8:[\"locale\",\"fr\",\"d\"]\n9:[\"slug\",\"articles-erreurs_javascript\",\"oc\"]\nc:{\"fontFamily\":\"system-ui,\\\"Segoe UI\\\",Roboto,Helvetica,Arial,sans-serif,\\\"Apple Color Emoji\\\",\\\"Segoe UI Emoji\\\"\",\"height\":\"100vh\",\"textAlign\":\"center\",\"display\":\"flex\",\"flexDirection\":\"column\",\"alignItems\":\"center\",\"justifyContent\":\"center\"}\nd:{\"display\":\"inline-block\",\"margin\":\"0 20px 0 0\",\"padding\":\"0 23px 0 0\",\"fontSize\":24,\"fontWeight\":500,\"verticalAlign\":\"top\",\"lineHei"])</script><script>self.__next_f.push([1,"ght\":\"49px\"}\ne:{\"display\":\"inline-block\"}\nf:{\"fontSize\":14,\"fontWeight\":400,\"lineHeight\":\"49px\",\"margin\":0}\n"])</script><script>self.__next_f.push([1,"1:[null,[\"$\",\"$L4\",null,{\"buildId\":\"0iMe3vHKCcHPUIZbS85nF\",\"assetPrefix\":\"https://insertafter.com\",\"initialCanonicalUrl\":\"/fr/articles-erreurs_javascript\",\"initialTree\":[\"\",{\"children\":[\"(withLocale)\",{\"children\":[[\"locale\",\"fr\",\"d\"],{\"children\":[[\"slug\",\"articles-erreurs_javascript\",\"oc\"],{\"children\":[\"__PAGE__?{\\\"locale\\\":\\\"fr\\\",\\\"slug\\\":[\\\"articles-erreurs_javascript\\\"]}\",{}]}]},\"$undefined\",\"$undefined\",true]}]}],\"initialSeedData\":[\"\",{\"children\":[\"(withLocale)\",{\"children\":[[\"locale\",\"fr\",\"d\"],{\"children\":[[\"slug\",\"articles-erreurs_javascript\",\"oc\"],{\"children\":[\"__PAGE__\",{},[\"$L5\",\"$L6\",null]]},[\"$\",\"$L7\",null,{\"parallelRouterKey\":\"children\",\"segmentPath\":[\"children\",\"(withLocale)\",\"children\",\"$8\",\"children\",\"$9\",\"children\"],\"loading\":\"$undefined\",\"loadingStyles\":\"$undefined\",\"loadingScripts\":\"$undefined\",\"hasLoading\":false,\"error\":\"$undefined\",\"errorStyles\":\"$undefined\",\"errorScripts\":\"$undefined\",\"template\":[\"$\",\"$La\",null,{}],\"templateStyles\":\"$undefined\",\"templateScripts\":\"$undefined\",\"notFound\":\"$undefined\",\"notFoundStyles\":\"$undefined\",\"styles\":[[\"$\",\"link\",\"0\",{\"rel\":\"stylesheet\",\"href\":\"https://insertafter.com/_next/static/css/cd8d1ff5c688cf4d.css\",\"precedence\":\"next\",\"crossOrigin\":\"\"}]]}]]},[null,\"$Lb\",null]]},[\"$\",\"$L7\",null,{\"parallelRouterKey\":\"children\",\"segmentPath\":[\"children\",\"(withLocale)\",\"children\"],\"loading\":\"$undefined\",\"loadingStyles\":\"$undefined\",\"loadingScripts\":\"$undefined\",\"hasLoading\":false,\"error\":\"$undefined\",\"errorStyles\":\"$undefined\",\"errorScripts\":\"$undefined\",\"template\":[\"$\",\"$La\",null,{}],\"templateStyles\":\"$undefined\",\"templateScripts\":\"$undefined\",\"notFound\":[[\"$\",\"title\",null,{\"children\":\"404: This page could not be found.\"}],[\"$\",\"div\",null,{\"style\":{\"fontFamily\":\"system-ui,\\\"Segoe UI\\\",Roboto,Helvetica,Arial,sans-serif,\\\"Apple Color Emoji\\\",\\\"Segoe UI Emoji\\\"\",\"height\":\"100vh\",\"textAlign\":\"center\",\"display\":\"flex\",\"flexDirection\":\"column\",\"alignItems\":\"center\",\"justifyContent\":\"center\"},\"children\":[\"$\",\"div\",null,{\"children\":[[\"$\",\"style\",null,{\"dangerouslySetInnerHTML\":{\"__html\":\"body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}\"}}],[\"$\",\"h1\",null,{\"className\":\"next-error-h1\",\"style\":{\"display\":\"inline-block\",\"margin\":\"0 20px 0 0\",\"padding\":\"0 23px 0 0\",\"fontSize\":24,\"fontWeight\":500,\"verticalAlign\":\"top\",\"lineHeight\":\"49px\"},\"children\":\"404\"}],[\"$\",\"div\",null,{\"style\":{\"display\":\"inline-block\"},\"children\":[\"$\",\"h2\",null,{\"style\":{\"fontSize\":14,\"fontWeight\":400,\"lineHeight\":\"49px\",\"margin\":0},\"children\":\"This page could not be found.\"}]}]]}]}]],\"notFoundStyles\":[],\"styles\":[[\"$\",\"link\",\"0\",{\"rel\":\"stylesheet\",\"href\":\"https://insertafter.com/_next/static/css/655a273f6fe37552.css\",\"precedence\":\"next\",\"crossOrigin\":\"\"}]]}]]},[\"$\",\"$L7\",null,{\"parallelRouterKey\":\"children\",\"segmentPath\":[\"children\"],\"loading\":\"$undefined\",\"loadingStyles\":\"$undefined\",\"loadingScripts\":\"$undefined\",\"hasLoading\":false,\"error\":\"$undefined\",\"errorStyles\":\"$undefined\",\"errorScripts\":\"$undefined\",\"template\":[\"$\",\"$La\",null,{}],\"templateStyles\":\"$undefined\",\"templateScripts\":\"$undefined\",\"notFound\":[[\"$\",\"title\",null,{\"children\":\"404: This page could not be found.\"}],[\"$\",\"div\",null,{\"style\":\"$c\",\"children\":[\"$\",\"div\",null,{\"children\":[[\"$\",\"style\",null,{\"dangerouslySetInnerHTML\":{\"__html\":\"body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}\"}}],[\"$\",\"h1\",null,{\"className\":\"next-error-h1\",\"style\":\"$d\",\"children\":\"404\"}],[\"$\",\"div\",null,{\"style\":\"$e\",\"children\":[\"$\",\"h2\",null,{\"style\":\"$f\",\"children\":\"This page could not be found.\"}]}]]}]}]],\"notFoundStyles\":[],\"styles\":null}]],\"initialHead\":[false,\"$L10\"],\"globalErrorComponent\":\"$11\"}]]\n"])</script><script>self.__next_f.push([1,"12:\"$Sreact.strict_mode\"\n13:I[4630,[\"176\",\"static/chunks/176-0a03be62efe5672b.js\",\"960\",\"static/chunks/app/(withLocale)/%5Blocale%5D/layout-3730889d272dc34b.js\"],\"\"]\n14:I[4341,[\"176\",\"static/chunks/176-0a03be62efe5672b.js\",\"960\",\"static/chunks/app/(withLocale)/%5Blocale%5D/layout-3730889d272dc34b.js\"],\"\"]\n"])</script><script>self.__next_f.push([1,"b:[\"$\",\"$12\",null,{\"children\":[\"$\",\"html\",null,{\"lang\":\"fr\",\"children\":[\"$\",\"body\",null,{\"className\":\"layout_body__jFZuu\",\"children\":[null,[\"$\",\"header\",null,{\"className\":\"header_iaHeader__4vCMA\",\"children\":[\"$\",\"div\",null,{\"className\":\"$undefined\",\"children\":[\"$\",\"h1\",null,{\"children\":[\"$\",\"a\",null,{\"href\":\"/fr\",\"title\":\"Retour Ã  lâ€™accueil\",\"className\":\"header_iaHeaderLogo__0uNr1\",\"children\":[\"$\",\"img\",null,{\"src\":\"/images/logo.svg\",\"alt\":\"Logo dâ€™InsertAfter\"}]}]}]}]}],[\"$\",\"$L13\",null,{\"locale\":\"fr\",\"dictionary\":{\"home\":{\"label\":\"Accueil\",\"title\":\"Retour Ã  lâ€™accueil\"},\"blog\":{\"label\":\"Blog\",\"title\":\"Retour Ã  lâ€™index du blog\"},\"projects\":{\"label\":\"Projets\",\"title\":\"DÃ©couvrir mes divers projets\"},\"about\":{\"label\":\"Ã€ propos\",\"title\":\"En savoir plus Ã  mon sujet\"}}}],[\"$\",\"$L14\",null,{\"locale\":\"fr\"}],[\"$\",\"main\",null,{\"className\":\"layout_iaMain__6s2gl\",\"children\":[\"$\",\"div\",null,{\"children\":[\"$\",\"$L7\",null,{\"parallelRouterKey\":\"children\",\"segmentPath\":[\"children\",\"(withLocale)\",\"children\",\"$8\",\"children\"],\"loading\":\"$undefined\",\"loadingStyles\":\"$undefined\",\"loadingScripts\":\"$undefined\",\"hasLoading\":false,\"error\":\"$undefined\",\"errorStyles\":\"$undefined\",\"errorScripts\":\"$undefined\",\"template\":[\"$\",\"$La\",null,{}],\"templateStyles\":\"$undefined\",\"templateScripts\":\"$undefined\",\"notFound\":\"$L15\",\"notFoundStyles\":[[\"$\",\"link\",\"0\",{\"rel\":\"stylesheet\",\"href\":\"https://insertafter.com/_next/static/css/719d59f1b635ba1d.css\",\"precedence\":\"next\",\"crossOrigin\":\"\"}]],\"styles\":null}]}]}],[\"$\",\"aside\",null,{\"className\":\"social_iaSocial__kdR1x\",\"children\":[\"$\",\"nav\",null,{\"children\":[\"$\",\"ul\",null,{\"children\":[[\"$\",\"li\",null,{\"children\":[\"$\",\"a\",null,{\"href\":\"https://mastodon.social/@nfroidure\",\"rel\":\"me\",\"title\":\"Me suivre sur Mastodon\",\"className\":\"social_iaSocialMastodon__MTF6o\",\"children\":[\"$\",\"span\",null,{\"children\":\"Mastodon\"}]}]}],[\"$\",\"li\",null,{\"children\":[\"$\",\"a\",null,{\"href\":\"https://bsky.app/profile/nfroidure.bsky.social\",\"rel\":\"me\",\"title\":\"Me suivre sur Bluesky\",\"className\":\"social_iaSocialBluesky__fKW6M\",\"children\":[\"$\",\"span\",null,{\"children\":\"Bluesky\"}]}]}],[\"$\",\"li\",null,{\"children\":[\"$\",\"a\",null,{\"href\":\"https://twitter.com/nfroidure\",\"title\":\"Me suivre sur Twitter\",\"className\":\"social_iaSocialTwitter__9nU6R\",\"children\":[\"$\",\"span\",null,{\"children\":\"Twitter\"}]}]}],[\"$\",\"li\",null,{\"children\":[\"$\",\"a\",null,{\"href\":\"https://github.com/nfroidure\",\"title\":\"Voir mes projets sur GitHub\",\"className\":\"social_iaSocialGithub__vcU5_\",\"children\":[\"$\",\"span\",null,{\"children\":\"GitHub\"}]}]}],[\"$\",\"li\",null,{\"children\":[\"$\",\"a\",null,{\"href\":\"https://www.npmjs.org/~nfroidure\",\"title\":\"Voir mon compte NPM\",\"className\":\"social_iaSocialNpm__6nku5\",\"children\":[\"$\",\"span\",null,{\"children\":\"NPM\"}]}]}],[\"$\",\"li\",null,{\"children\":[\"$\",\"a\",null,{\"href\":\"https://www.linkedin.com/in/nfroidure/\",\"rel\":\"me\",\"title\":\"Voir mon profil LinkedIn\",\"className\":\"social_iaSocialLinkedin__pWACO\",\"children\":[\"$\",\"span\",null,{\"children\":\"LinkedIn\"}]}]}],[\"$\",\"li\",null,{\"children\":[\"$\",\"a\",null,{\"href\":\"https://facebook.com/nicolas.froidure.douai\",\"rel\":\"me\",\"title\":\"Me suivre sur Facebook\",\"className\":\"social_iaSocialFacebook__w4wfH\",\"children\":[\"$\",\"span\",null,{\"children\":\"Facebook\"}]}]}],[\"$\",\"li\",null,{\"children\":[\"$\",\"a\",null,{\"href\":\"/fr/blog/index.atom\",\"title\":\"Sâ€™abonner au flux RSS\",\"className\":\"social_iaSocialFeed__gEQQ0\",\"children\":[\"$\",\"span\",null,{\"children\":\"Flux RSS\"}]}]}]]}]}]}],[\"$\",\"footer\",null,{\"className\":\"footer_iaFooter__rhlnQ\",\"children\":[\"$\",\"p\",null,{\"className\":\"footer_iaFooterContent__VZndF\",\"children\":[\"Â© Nicolas Froidure 2012-\",2024]}]}]]}]}]}]\n"])</script><script>self.__next_f.push([1,"15:[\"$\",\"section\",null,{\"className\":\"contentBlock_root__802UK\",\"children\":[[\"$\",\"h1\",null,{\"className\":\"h1_root__VPXQG\",\"children\":\"Page not found\"}],[\"$\",\"p\",null,{\"className\":\"not-found_p__fvcGn\",\"children\":[[\"$\",\"img\",null,{\"className\":\"not-found_img__IT2jY\",\"src\":\"https://media.giphy.com/media/12mPcp41D9a1i0/giphy.gif\",\"alt\":\"404.png\"}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":\"It seems that the page you look for doesn't exist.\"}]]}]]}]\n"])</script><script>self.__next_f.push([1,"16:I[5250,[\"176\",\"static/chunks/176-0a03be62efe5672b.js\",\"790\",\"static/chunks/790-a194f74901d33197.js\",\"730\",\"static/chunks/app/(withLocale)/%5Blocale%5D/%5B%5B...slug%5D%5D/page-992952bac383cfc2.js\"],\"\"]\n"])</script><script>self.__next_f.push([1,"6:[[\"$\",\"section\",null,{\"className\":\"contentBlock_root__802UK\",\"children\":[[[\"$\",\"h1\",null,{\"className\":\"h1_root__VPXQG\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Gestion des erreurs en JavaScript\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"GÃ©rer les erreurs en JavaScript peut paraÃ®tre simple, mais il n'est pas toujours Ã©vident de choisir entre exceptions, retours d'erreurs et callbacks d'erreurs.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Le dernier Ch'tiJS a dÃ©cidÃ©ment Ã©tÃ© la source de nombreuses rÃ©flexions. L'une d'entre elle concernait les erreurs. Tout est parti d'une assertion simpleÂ : les promises doivent-elles attraper les exceptions et les transformer en erreurs rÃ©cupÃ©rables dans un then ou les laisser remonter la pile d'exÃ©cutionÂ ?\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"C'est que fait, par exemple, RSVP.js. La logique du resolver est englobÃ©e dans \"}],[\"$\",\"$L16\",null,{\"href\":\"https://github.com/tildeio/rsvp.js/blob/277eb5eab912a9c3df8228a2647fcab84374615a/lib/rsvp/promise.js#L47\",\"className\":\"a_root__WEPS7\",\"title\":\"Voir le code en question\",\"target\":\"_blank\",\"children\":[null,[[\"$\",\"span\",\"0\",{\"children\":\"un bloc try...catch\"}]]]}],[\"$\",\"span\",\"2\",{\"children\":\" qui transforme toute exception en erreur qui fait Ã©chouer la promise. Dans \"}],[\"$\",\"$L16\",null,{\"href\":\"promises%5Fjavascript%5Fouvrons%5Fcapot.html\",\"className\":\"a_root__WEPS7\",\"title\":\"Voir l'article concernant cette implÃ©mentation\",\"target\":\"_self\",\"children\":[null,[[\"$\",\"span\",\"0\",{\"children\":\"mon implÃ©mentation des promises\"}]]]}],[\"$\",\"span\",\"4\",{\"children\":\", j'ai choisi de ne pas utiliser cette approche afin de pouvoir dÃ©cider, au cas par cas, d'attraper ou non une exception envoyÃ©e par le resolver d'une promise.\"}]]}],[\"$\",\"h2\",null,{\"className\":\"h2_root__Y02lX\",\"children\":[\"$\",\"span\",null,{\"className\":\"anchored_root__HWi6K\",\"children\":[[[\"$\",\"span\",\"0\",{\"children\":\"Typologie des erreurs\"}]],\" \",[\"$\",\"small\",null,{\"children\":[\"$\",\"$L16\",null,{\"href\":\"#typologie-des-erreurs\",\"className\":\"anchored_icon__N5Sem\",\"id\":\"typologie-des-erreurs\",\"title\":\"Lien vers cette section\",\"children\":[\"$\",\"span\",null,{\"children\":\"ğŸ”—\"}]}]}]]}]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Avant d'aller plus loin, je voudrais revenir sur la typologie des erreurs que l'on peut rencontrer au cours de l'exÃ©cution d'un programme. Pour moi, il existe trois classes d'erreursÂ :\"}]]}],[\"$\",\"ul\",null,{\"className\":\"ul_ul__qh0e9\",\"children\":[[\"$\",\"li\",null,{\"className\":\"li_root__oY9Dv\",\"children\":[[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"les erreurs fatalesÂ : ce sont des erreurs qui rÃ©sultent d'une erreur de programmation. Ces derniÃ¨res ont la particularitÃ© de ne pas avoir Ã©tÃ© prÃ©vues par le dÃ©veloppeur puisqu'elles rÃ©sultent d'un bug. Elles laissent donc le programme dans un Ã©tat indÃ©terminÃ©. On pourrait comparer cette erreur aux erreurs HTTP 500+ ;\"}]]}]]}],[\"$\",\"li\",null,{\"className\":\"li_root__oY9Dv\",\"children\":[[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"les erreurs d'utilisationÂ : elles sont causÃ©es par une mauvaise utilisation du programme (mauvaise saisies de l'utilisateur principalement) ;\"}]]}]]}],[\"$\",\"li\",null,{\"className\":\"li_root__oY9Dv\",\"children\":[[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"les erreurs d'interactionÂ : un programme interagit frÃ©quemment avec d'autres composants (le rÃ©seau, le matÃ©riel etc...). Il arrive qu'une interaction se termine de faÃ§on imprÃ©vue ou incohÃ©rente. Un programme se doit d'anticiper ces erreurs et de les gÃ©rer au mieux.\"}]]}]]}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Attention, certaines erreurs d'interaction peuvent Ãªtre dues Ã  une erreur fatale dans un programme avec lequel votre code interagit. C'est le cas quand une requÃªte XHR renvoie une erreur 500. Il faut donc veiller Ã  ne pas confondre erreurs d'interactions et erreurs fatales. Pour ce faire, il faut bien dÃ©limiter les erreurs qui sont dues Ã  votre code.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"On pourrait considÃ©rer que les erreurs d'utilisation sont des erreurs d'interaction avec l'utilisateur. D'ailleurs, dans l'absolu, l'interaction avec des systÃ¨me tiers peut aussi induire des entrÃ©es qui nÃ©cessitent d'Ãªtre filtrÃ©es de la mÃªme faÃ§on que les entrÃ©es utilisateur. Je fais nÃ©anmoins la diffÃ©rence ici car la rÃ©ponse Ã  ces erreurs sera lÃ©gÃ¨rement diffÃ©rente.\"}]]}],[\"$\",\"h2\",null,{\"className\":\"h2_root__Y02lX\",\"children\":[\"$\",\"span\",null,{\"className\":\"anchored_root__HWi6K\",\"children\":[[[\"$\",\"span\",\"0\",{\"children\":\"La gestion d'erreur\"}]],\" \",[\"$\",\"small\",null,{\"children\":[\"$\",\"$L16\",null,{\"href\":\"#la-gestion-d-erreur\",\"className\":\"anchored_icon__N5Sem\",\"id\":\"la-gestion-d-erreur\",\"title\":\"Lien vers cette section\",\"children\":[\"$\",\"span\",null,{\"children\":\"ğŸ”—\"}]}]}]]}]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"J'imagine que ces types vous Ã©voquent certaines stratÃ©gies pour leur gestion. En JavaScript, on compte divers type de gestion d'erreurÂ :\"}]]}],[\"$\",\"ul\",null,{\"className\":\"ul_ul__qh0e9\",\"children\":[[\"$\",\"li\",null,{\"className\":\"li_root__oY9Dv\",\"children\":[[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"les exceptionsÂ : elles permettent Ã  une erreur de remonter la pile d'exÃ©cution. On peut les attraper Ã  tout moment Ã  l'aide d'un bloc \"}],[\"$\",\"code\",null,{\"className\":\"code_root__5JLV_\",\"children\":\"try...catch\"}],[\"$\",\"span\",\"2\",{\"children\":\". Si elles ne sont pas attrapÃ©es, les exceptions finissent par interrompre l'execution de votre JavaScript (plus prÃ©cisÃ©ment de la pile d'exÃ©cution JavaScript).\"}]]}]]}],[\"$\",\"li\",null,{\"className\":\"li_root__oY9Dv\",\"children\":[[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"les retours d'erreurÂ : il s'agit d'exploiter le retour de fonction pour indiquer qu'une erreur a eu lieu. On peut considÃ©rer que si \"}],[\"$\",\"code\",null,{\"className\":\"code_root__5JLV_\",\"children\":\"parseInt\"}],[\"$\",\"span\",\"2\",{\"children\":\" retourne \"}],[\"$\",\"code\",null,{\"className\":\"code_root__5JLV_\",\"children\":\"NaN\"}],[\"$\",\"span\",\"4\",{\"children\":\" alors on a affaire Ã  une erreur. Cette approche est trÃ¨s utilisÃ©e en C pour les appels systÃ¨mes.\"}]]}]]}],[\"$\",\"li\",null,{\"className\":\"li_root__oY9Dv\",\"children\":[[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"l'approche via une variable globaleÂ : trÃ¨s prÃ©sente en C Ã©galement avec la variable \"}],[\"$\",\"code\",null,{\"className\":\"code_root__5JLV_\",\"children\":\"errno\"}],[\"$\",\"span\",\"2\",{\"children\":\". En JavaScript, on Ã©vite les variables globales, on utilisera donc pas cette approche.\"}]]}]]}],[\"$\",\"li\",null,{\"className\":\"li_root__oY9Dv\",\"children\":[[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"l'approche via callback d'erreurÂ : ils sont principalement utilisÃ©s pour les opÃ©rations asynchrones, bien que rien n'empÃªche leur utilisation dans du code synchrone. C'est aussi l'approche qui est utilisÃ©e pour les promises. On passe un callback d'erreur Ã  la mÃ©thode then. La diffÃ©rence est que l'on passe autant de callbacks d'erreur qu'on le souhaite.\"}]]}]]}]]}],[\"$\",\"h2\",null,{\"className\":\"h2_root__Y02lX\",\"children\":[\"$\",\"span\",null,{\"className\":\"anchored_root__HWi6K\",\"children\":[[[\"$\",\"span\",\"0\",{\"children\":\"Comment gÃ©rer chaque type d'erreurÂ ?\"}]],\" \",[\"$\",\"small\",null,{\"children\":[\"$\",\"$L16\",null,{\"href\":\"#comment-gerer-chaque-type-d-erreur\",\"className\":\"anchored_icon__N5Sem\",\"id\":\"comment-gerer-chaque-type-d-erreur\",\"title\":\"Lien vers cette section\",\"children\":[\"$\",\"span\",null,{\"children\":\"ğŸ”—\"}]}]}]]}]}],[\"$\",\"h3\",null,{\"className\":\"h3_root__kVaqq\",\"children\":[\"$\",\"span\",null,{\"className\":\"anchored_root__HWi6K\",\"children\":[[[\"$\",\"span\",\"0\",{\"children\":\"Les erreurs fatales\"}]],\" \",[\"$\",\"small\",null,{\"children\":[\"$\",\"$L16\",null,{\"href\":\"#les-erreurs-fatales\",\"className\":\"anchored_icon__N5Sem\",\"id\":\"les-erreurs-fatales\",\"title\":\"Lien vers cette section\",\"children\":[\"$\",\"span\",null,{\"children\":\"ğŸ”—\"}]}]}]]}]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Les exceptions sont l'arme ultime pour gÃ©rer les erreurs fatales. En effet, puisque les erreurs fatales sont des erreurs de programmation, continuer l'exÃ©cution du programme est risquÃ©. Cela pourrait avoir des rÃ©percussions dÃ©sastreuses.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"JavaScript utilise intensivement les exceptions. Les erreurs de syntaxe, les mauvaises utilisations de son API native sont signalÃ©es dans la grande majoritÃ© des cas par des exceptions. C'est pourquoi je vous enjoins Ã  fait de mÃªme.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"La plupart du temps, il s'agit de lever une exception si l'une de vos fonctions/l'un de vos constructeurs reÃ§oit en paramÃ¨tre des arguments erronÃ©sÂ :\"}]]}],[\"$\",\"div\",null,{\"className\":\"code_root__5JLV_\",\"children\":[\"$\",\"code\",null,{\"children\":[\"$\",\"pre\",null,{\"children\":\"function repeat(n) {\\n  if (n \u003c 0) throw new Error(\\\"Bad n value.\\\");\\n  //...\\n}\"}]}]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Demander une rÃ©pÃ©tition d'une opÃ©ration avec une nombre nÃ©gatif de rÃ©pÃ©tition n'a pas de sens. Lancer une exception permet d'interrompre l'exÃ©cution et de permettre de dÃ©boguer facilement l'erreur dans le contexte oÃ¹ celle-ci est apparue.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"En effet, tous les outils pour dÃ©veloppeurs permettent de placer un point d'arrÃªt Ã  la levÃ©e d'une exception afin de permettre de visualiser les variables Ã  portÃ©es, la pile d'exÃ©cution et bien d'autres informations trÃ¨s utiles.\"}]]}],[\"$\",\"h3\",null,{\"className\":\"h3_root__kVaqq\",\"children\":[\"$\",\"span\",null,{\"className\":\"anchored_root__HWi6K\",\"children\":[[[\"$\",\"span\",\"0\",{\"children\":\"Les erreurs d'utilisation\"}]],\" \",[\"$\",\"small\",null,{\"children\":[\"$\",\"$L16\",null,{\"href\":\"#les-erreurs-d-utilisation\",\"className\":\"anchored_icon__N5Sem\",\"id\":\"les-erreurs-d-utilisation\",\"title\":\"Lien vers cette section\",\"children\":[\"$\",\"span\",null,{\"children\":\"ğŸ”—\"}]}]}]]}]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Les erreurs d'utilisation sont dues Ã  l'utilisateur. La logique veut donc que l'on traite ces erreurs en signalant celles-cis Ã  l'utilisateur. C'est donc une rÃ©ponse par le biais de l'interface utilisateur qui sera privilÃ©giÃ©e.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Cependant, pour activer cette rÃ©ponse, il faut en premier lieu la dÃ©tecter. Cette dÃ©tection se fait indiffÃ©remment par retour de fonction ou par callback d'erreur. D'ailleurs il est trÃ¨s possible que vous deviez mixer ces deux techniques pour bÃ¢tir une propagation d'erreur dans la pile d'exÃ©cution et annuler certaines actions suite Ã  une erreur.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"L'idÃ©e Ã  retenir est d'Ã©viter Ã  tout prix de gÃ©rer les erreurs des utilisateurs Ã  l'aide d'exceptions. En effet, vous prenez le risque de mettre fin Ã  l'exÃ©cution pour une erreur normale dans le cycle de vie d'une application. Les utilisateurs se trompent souvent.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"On prendra donc soin de filtrer un entier obtenu grÃ¢ce Ã  une saisie de l'utilisateur avant de l'utiliser comme argument de la fonction \"}],[\"$\",\"code\",null,{\"className\":\"code_root__5JLV_\",\"children\":\"repeat\"}],[\"$\",\"span\",\"2\",{\"children\":\" dÃ©finie ci-dessus.\"}]]}],[\"$\",\"h3\",null,{\"className\":\"h3_root__kVaqq\",\"children\":[\"$\",\"span\",null,{\"className\":\"anchored_root__HWi6K\",\"children\":[[[\"$\",\"span\",\"0\",{\"children\":\"Les erreurs d'interaction\"}]],\" \",[\"$\",\"small\",null,{\"children\":[\"$\",\"$L16\",null,{\"href\":\"#les-erreurs-d-interaction\",\"className\":\"anchored_icon__N5Sem\",\"id\":\"les-erreurs-d-interaction\",\"title\":\"Lien vers cette section\",\"children\":[\"$\",\"span\",null,{\"children\":\"ğŸ”—\"}]}]}]]}]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Lorsque votre code interagit avec des Ã©lÃ©ments extÃ©rieurs, qu'il s'agisse de serveurs distant ou du hardware, le bon sens veut que cela se fasse de maniÃ¨re asynchrone. Les callbacks d'erreur sont donc indiquÃ©s pour remplir le job. C'est d'ailleurs de cette faÃ§on que l'API JavaScript vous informe sur le succÃ¨s ou l'Ã©chec d'une opÃ©ration asynchrone.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Ce n'est malheureusement pas toujours le cas. Par exemple, une requÃªte XHR synchrone qui Ã©choue Ã  cause d'un problÃ¨me rÃ©seau lance une exception. Dans ce cas prÃ©cis, mieux vaut Ãªtre au courant afin de placer le bloc \"}],[\"$\",\"code\",null,{\"className\":\"code_root__5JLV_\",\"children\":\"try...catch\"}],[\"$\",\"span\",\"2\",{\"children\":\" qui vous permettra de gÃ©rer ce cas sans mettre en pÃ©ril l'exÃ©cution de votre JavaScript.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Vous vous dÃ®tes peut-Ãªtre qu'il peut arriver qu'une erreur de programmation mÃ¨ne Ã  la mauvaise utilisation d'une API asynchrone et que dans ce cas, un callback d'erreur sera appelÃ© pour une erreur fatale. Ce que j'ai totalement contre indiquÃ© tout Ã  l'heure.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Pour le cas d'une API bien pensÃ©e, cela ne devrait pas Ãªtre le cas. Une API asynchrone doit dÃ©tecter toute mauvaise utilisation et lancer une exception au moment de son lancement. Par exemple, si vous lancez une requÃªte XHR vers une origine diffÃ©rente de votre site, c'est une erreur de programmation. Une exception doit Ãªtre lancÃ©e immÃ©diatement.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"L'astuce pour Ã©viter tout effet de bord est de lancer l'opÃ©ration asynchrone avant d'assigner les callbacks ou Ã©couteurs d'Ã©vÃ¨nements. De cette faÃ§on vous Ãªtes certains que ces derniers ne seront jamais appelÃ©s en cas de levÃ©e d'une exception.\"}]]}],[\"$\",\"h2\",null,{\"className\":\"h2_root__Y02lX\",\"children\":[\"$\",\"span\",null,{\"className\":\"anchored_root__HWi6K\",\"children\":[[[\"$\",\"span\",\"0\",{\"children\":\"Retour sur les promises\"}]],\" \",[\"$\",\"small\",null,{\"children\":[\"$\",\"$L16\",null,{\"href\":\"#retour-sur-les-promises\",\"className\":\"anchored_icon__N5Sem\",\"id\":\"retour-sur-les-promises\",\"title\":\"Lien vers cette section\",\"children\":[\"$\",\"span\",null,{\"children\":\"ğŸ”—\"}]}]}]]}]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Avec cette toile de fond en tÃªte, il apparaÃ®t donc Ã©vident que les exceptions ne doivent pas Ãªtre attrapÃ©es au sein des promises. En effet, au delÃ  de l'avantage d'une gestion au cas par cas, on voit aussi que cela pourrait compromettre la dÃ©tection d'une erreur fatale causant la continuation de l'exÃ©cution Ã  partir d'un Ã©tat non-prÃ©vu donc indÃ©terminÃ©.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"On pourrait penser Ã©pargner l'utilisateur d'erreurs fatales qui en dÃ©finitive mettent fin Ã  l'exÃ©cution, mais au final, on rend les erreurs de programmation plus difficiles Ã  dÃ©tecter et les bugs plus sournois.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Il faut comprendre que les erreurs fatales doivent Ãªtre exceptionnelles. Un programme qui arrive en production doit Ãªtre dÃ©nuÃ© d'erreurs fatales dans l'absolu. Malheureusement, dans le monde rÃ©el, ces erreurs peuvent survenir en production.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Devons nous laisser les utilisateurs face Ã  une application qui ne fonctionne plusÂ ? SÃ»rement pas. L'approche correcte pour gÃ©rer les erreurs fatales est finalement assez simple. Il suffit d'utiliser un callback d'erreur gÃ©nÃ©rique, d'informer l'utilisateur d'une erreur anormale et de redÃ©marrer l'application. On peut aussi imaginer rÃ©cupÃ©rer ces erreurs fatales grÃ¢ce Ã  une requÃªte XHR et de crÃ©er directement une issue sur GitHub. Un simple exempleÂ :\"}]]}],[\"$\",\"div\",null,{\"className\":\"code_root__5JLV_\",\"children\":[\"$\",\"code\",null,{\"children\":[\"$\",\"pre\",null,{\"children\":\"window.onerror = function (msg, url, line) {\\n  //demande de dÃ©tails\\n  var details = prompt(\\n    \\\"Une erreur fatale a eu lieu, vous pouvez donner quelques dÃ©tails pour aider le webmasterÂ :\\\",\\n  );\\n  // envoi des dÃ©tails via XHR puis rafraÃ®chissement de la page\\n  document.location.refresh();\\n};\"}]}]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Ici, j'ai volontairement simplifiÃ©. Dans la pratique, on placera un timeout pour laisser le temps Ã  toutes les erreurs consÃ©cutives d'Ãªtre levÃ©es avant de demander du feedback et de redÃ©marrer l'application. On peut aussi donner le temps Ã  l'utilisateur de rÃ©cupÃ©rer certains contenus (un formulaire Ã  moitiÃ© rempli, par exemple) avant le rafraÃ®chissement.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"L'avantage de cette approche est qu'on informe l'utilisateur d'un problÃ¨me plutÃ´t que vouloir le lui cacher quitte Ã  ce que ce dernier provoque un effet papillon plus difficile Ã  dÃ©tecter.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"VoilÃ , j'espÃ¨re que cet article vous a donnÃ© envie de gÃ©rer les erreurs de vos applications. Je vous propose de lire ce billet sur l'\"}],[\"$\",\"$L16\",null,{\"href\":\"bugmeback.html\",\"className\":\"a_root__WEPS7\",\"title\":\"Voir le billet sur BugMeBack\",\"target\":\"_self\",\"children\":[null,[[\"$\",\"span\",\"0\",{\"children\":\"extension BugMeBack\"}]]]}],[\"$\",\"span\",\"2\",{\"children\":\" qui propose une gestion intÃ©ressante des erreurs pour vos applications.\"}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"span\",\"0\",{\"children\":\"Petit ajoutÂ : J'ai Ã©crit \"}],[\"$\",\"$L16\",null,{\"href\":\"erreurs%5Fjavascript%5Fameliorees.html\",\"className\":\"a_root__WEPS7\",\"title\":\"Lire l'article\",\"target\":\"_self\",\"children\":[null,[[\"$\",\"span\",\"0\",{\"children\":\"un autre article sur les erreurs en JS\"}]]]}],[\"$\",\"span\",\"2\",{\"children\":\" mais cette fois avec des exemples concrets.\"}]]}]],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[\"Published at\",\" \",\"mardi 30 juillet 2013 Ã  15:18:42\",\".\"]}]]}],[\"$\",\"section\",null,{\"className\":\"contentBlock_root__802UK\",\"children\":[\"$\",\"aside\",null,{\"className\":\"share_root__Qm_RR\",\"children\":[[\"$\",\"h2\",null,{\"className\":\"h2_root__Y02lX\",\"children\":\"Commenter et partager\"}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[[\"$\",\"$L16\",null,{\"href\":\"https://www.facebook.com/sharer/sharer.php?u=https%3A%2F%2Finsertafter.com%2Fblog%2Ferreurs_javascript\",\"className\":\"a_root__WEPS7\",\"title\":\"Commenter sur Facebook\",\"target\":\"_blank\",\"children\":[null,[[\"$\",\"span\",null,{\"className\":\"share_icon__bi4Q3 share_facebook__JKn8d\"}],\"Facebook\"]]}],\" - \",[\"$\",\"$L16\",null,{\"href\":\"https://twitter.com/intent/tweet/?url=https%3A%2F%2Finsertafter.com%2Fblog%2Ferreurs_javascript\u0026text=Gestion%20des%20erreurs%20en%20JavaScript\u0026via=nfroidure\",\"className\":\"a_root__WEPS7\",\"title\":\"Commenter sur Twitter\",\"target\":\"_blank\",\"children\":[null,[[\"$\",\"span\",null,{\"className\":\"share_icon__bi4Q3 \"}],\"Twitter\"]]}],\" - \",[\"$\",\"$L16\",null,{\"href\":\"mailto:nicolas.froidure@gmail.com\",\"className\":\"a_root__WEPS7\",\"title\":\"Me rÃ©pondre par mail\",\"target\":\"_self\",\"children\":[null,[[\"$\",\"span\",null,{\"className\":\"share_icon__bi4Q3 share_mail__yvBDK\"}],\"Courriel\"]]}]]}]]}]}],[\"$\",\"section\",null,{\"className\":\"contentBlock_root__802UK\",\"children\":[\"$\",\"aside\",null,{\"className\":\"page_linkedEntries__jiPuT\",\"children\":[[\"$\",\"h2\",null,{\"className\":\"h2_root__Y02lX\",\"children\":\"Articles similaires\"}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W\",\"children\":[\"Dans les catÃ©gories\",\" \",\"JavaScript et Gestion des erreurs\",\".\"]}],[\"$\",\"div\",null,{\"className\":\"items_entries__5hkLG\",\"children\":[[\"$\",\"div\",\"choisir_module_nodejs\",{\"className\":\"items_entry_item__GWeg7\",\"children\":[null,[\"$\",\"h2\",null,{\"className\":\"h2_root__Y02lX items_entry_title__vJ_8h\",\"children\":[\"$\",\"$L16\",null,{\"href\":\".//choisir_module_nodejs\",\"className\":\"a_root__WEPS7 items_entry_link__EtVod\",\"title\":\"Lire cet article de blog\",\"target\":\"_self\",\"children\":[null,\"Bien choisir ses modules NodeJSÂ : ma mÃ©thode\"]}]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W items_entry_description__ZWt9y\",\"children\":[\"Hier Ã  la troisiÃ¨me Ã©dition de ChtiJS, un question a Ã©tÃ© soulevÃ©e sur la confiance que l'on pouvait avoir dans les modules NodeJS. Je vous dÃ©voile donc ma mÃ©thode pour trancher entre npm install et npm remove ;).\",\" \",[\"$\",\"br\",null,{}],[\"$\",\"$L16\",null,{\"href\":\".//choisir_module_nodejs\",\"className\":\"a_root__WEPS7\",\"title\":\"Lire cet article de blog\",\"target\":\"_self\",\"children\":[null,\"Lire la suite\"]}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W items_entry_time__iNrot\",\"children\":[\"Published at\",\" \",\"vendredi 26 juillet 2013 Ã  16:04:25\",\".\"]}],[\"$\",\"div\",null,{\"className\":\"items_clear__3fyGk\"}]]}],[\"$\",\"div\",\"utiliser_requirejs\",{\"className\":\"items_entry_item__GWeg7\",\"children\":[null,[\"$\",\"h2\",null,{\"className\":\"h2_root__Y02lX items_entry_title__vJ_8h\",\"children\":[\"$\",\"$L16\",null,{\"href\":\".//utiliser_requirejs\",\"className\":\"a_root__WEPS7 items_entry_link__EtVod\",\"title\":\"Lire cet article de blog\",\"target\":\"_self\",\"children\":[null,\"Astuces pour utiliser pleinement RequireJS\"]}]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W items_entry_description__ZWt9y\",\"children\":[\"Pour un projet dont je vous reparlerai bientÃ´t, j'ai utilisÃ© RequireJS pour la premiÃ¨re fois en production. J'en profite pour vous faire un petit retour d'expÃ©rience avec quelques astuces de mon cru ;).\",\" \",[\"$\",\"br\",null,{}],[\"$\",\"$L16\",null,{\"href\":\".//utiliser_requirejs\",\"className\":\"a_root__WEPS7\",\"title\":\"Lire cet article de blog\",\"target\":\"_self\",\"children\":[null,\"Lire la suite\"]}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W items_entry_time__iNrot\",\"children\":[\"Published at\",\" \",\"jeudi 11 juillet 2013 Ã  11:57:45\",\".\"]}],[\"$\",\"div\",null,{\"className\":\"items_clear__3fyGk\"}]]}],[\"$\",\"div\",\"promises_javascript_ouvrons_capot\",{\"className\":\"items_entry_item__GWeg7\",\"children\":[null,[\"$\",\"h2\",null,{\"className\":\"h2_root__Y02lX items_entry_title__vJ_8h\",\"children\":[\"$\",\"$L16\",null,{\"href\":\".//promises_javascript_ouvrons_capot\",\"className\":\"a_root__WEPS7 items_entry_link__EtVod\",\"title\":\"Lire cet article de blog\",\"target\":\"_self\",\"children\":[null,\"Les promises avec JavaScriptÂ : ouvrons le capotÂ !\"]}]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W items_entry_description__ZWt9y\",\"children\":[\"Dans le petit monde de JavaScript les promises font l'actualitÃ©. Je vous propose de comprendre les promises par l'implÃ©mentation.\",\" \",[\"$\",\"br\",null,{}],[\"$\",\"$L16\",null,{\"href\":\".//promises_javascript_ouvrons_capot\",\"className\":\"a_root__WEPS7\",\"title\":\"Lire cet article de blog\",\"target\":\"_self\",\"children\":[null,\"Lire la suite\"]}]]}],[\"$\",\"p\",null,{\"className\":\"p_root__SHE_W items_entry_time__iNrot\",\"children\":[\"Published at\",\" \",\"samedi 8 juin 2013 Ã  12:04:17\",\".\"]}],[\"$\",\"div\",null,{\"className\":\"items_clear__3fyGk\"}]]}]]}]]}]}]]\n"])</script><script>self.__next_f.push([1,"10:[[\"$\",\"meta\",\"0\",{\"name\":\"viewport\",\"content\":\"width=device-width, initial-scale=1\"}],[\"$\",\"meta\",\"1\",{\"name\":\"theme-color\",\"content\":\"#3e78b2\"}],[\"$\",\"meta\",\"2\",{\"charSet\":\"utf-8\"}],[\"$\",\"title\",\"3\",{\"children\":\"Gestion des erreurs en JavaScript - Nicolas Froidure\"}],[\"$\",\"meta\",\"4\",{\"name\":\"description\",\"content\":\"GÃ©rer les erreurs en JavaScript peut paraÃ®tre simple, mais il n'est pas toujours Ã©vident de choisir entre exceptions, retours d'erreurs et callbacks d'erreurs.\"}],[\"$\",\"link\",\"5\",{\"rel\":\"author\",\"href\":\"https://insertafter.com\"}],[\"$\",\"meta\",\"6\",{\"name\":\"author\",\"content\":\"Nicolas Froidure\"}],[\"$\",\"meta\",\"7\",{\"name\":\"robots\",\"content\":\"index, follow\"}],[\"$\",\"link\",\"8\",{\"rel\":\"canonical\",\"href\":\"https://insertafter.com/fr/blog/erreurs_javascript\"}],[\"$\",\"meta\",\"9\",{\"property\":\"og:title\",\"content\":\"Gestion des erreurs en JavaScript - Nicolas Froidure\"}],[\"$\",\"meta\",\"10\",{\"property\":\"og:description\",\"content\":\"GÃ©rer les erreurs en JavaScript peut paraÃ®tre simple, mais il n'est pas toujours Ã©vident de choisir entre exceptions, retours d'erreurs et callbacks d'erreurs.\"}],[\"$\",\"meta\",\"11\",{\"property\":\"og:url\",\"content\":\"https://insertafter.com/fr/blog/erreurs_javascript\"}],[\"$\",\"meta\",\"12\",{\"property\":\"og:site_name\",\"content\":\"Nicolas Froidure\"}],[\"$\",\"meta\",\"13\",{\"property\":\"og:locale\",\"content\":\"fr_FR\"}],[\"$\",\"meta\",\"14\",{\"property\":\"og:image:alt\",\"content\":\"BanniÃ¨re du site\"}],[\"$\",\"meta\",\"15\",{\"property\":\"og:image\",\"content\":\"https://insertafter.com/images/banner.png\"}],[\"$\",\"meta\",\"16\",{\"property\":\"og:type\",\"content\":\"article\"}],[\"$\",\"meta\",\"17\",{\"name\":\"twitter:card\",\"content\":\"summary\"}],[\"$\",\"meta\",\"18\",{\"name\":\"twitter:site\",\"content\":\"@nfroidure\"}],[\"$\",\"meta\",\"19\",{\"name\":\"twitter:creator\",\"content\":\"@nfroidure\"}],[\"$\",\"meta\",\"20\",{\"name\":\"twitter:title\",\"content\":\"Gestion des erreurs en JavaScript - Nicolas Froidure\"}],[\"$\",\"meta\",\"21\",{\"name\":\"twitter:description\",\"content\":\"GÃ©rer les erreurs en JavaScript peut paraÃ®tre simple, mais il n'est pas toujours Ã©vident de choisir entre exceptions, retours d'erreurs et callbacks d'erreurs.\"}],[\"$\",\"meta\",\"22\",{\"name\":\"twitter:image:alt\",\"content\":\"BanniÃ¨re du site\"}],[\"$\",\"meta\",\"23\",{\"name\":\"twitter:image\",\"content\":\"https://insertafter.com/images/banner.png\"}],[\"$\",\"link\",\"24\",{\"rel\":\"icon\",\"href\":\"/images/favicon.svg\",\"type\":\"image/svg+xml\",\"sizes\":\"any\"}],[\"$\",\"link\",\"25\",{\"rel\":\"icon\",\"href\":\"/images/favicon-16.png\",\"type\":\"image/png\",\"sizes\":\"16x16\"}],[\"$\",\"link\",\"26\",{\"rel\":\"icon\",\"href\":\"/images/favicon-128.png\",\"type\":\"image/png\",\"sizes\":\"128x128\"}]]\n"])</script><script>self.__next_f.push([1,"5:null\n"])</script><script>self.__next_f.push([1,""])</script></body></html>